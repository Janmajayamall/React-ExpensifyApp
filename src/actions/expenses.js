import uuid from 'uuid';
import database from './../firebase/firebase.js';
import { get } from 'http';

//Actions for Expenses 
export const addExpense = (expense) => ({
    type:'ADD_EXPENSE',
    expense: expense
});

export const startAddExpense = (expenseData = {}) =>{
    return (dispatch, getState) =>{
        const uid = getState().auth.uid;
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
            } = expenseData;
   
   const expense = {description, note, amount, createdAt};
          
   return database.ref(`users/${uid}/expenses`).push(expense).then((ref)=>{
     dispatch(addExpense({
         id: ref.key,
         ...expense
     }));
   });
    };
};

export const removeExpense = (id) => ({
    type: 'REMOVE_EXPENSE',
    id: id
});

export const startRemoveExpense = (id) =>{
    return (dispatch, getState) =>{
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).remove().then(()=>{
            dispatch(removeExpense(id));
        });
    }
}

export const editExpense = (id, update)=>({
    type: 'EDIT_EXPENSE',
    id:id,
    update: update
});

export const startEditExpense = (id, update) =>{
    return (dispatch, getState) =>{
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update(update).then(()=>{
            dispatch(editExpense(id, update));
        });
    }
};



export const setExpenses = (expenses) =>({
    type: 'SET_EXPENSES',
    expenses: expenses
})

export const startSetExpenses = () => {
    return (dispatch, getState) =>{
        const uid = getState().auth.uid;
       return database.ref(`users/${uid}/expenses`).once("value").then((snapshot)=>{
            const expenses = [];

            snapshot.forEach((childSnapshot)=>{
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });

            dispatch(setExpenses(expenses));
        });
    };
};