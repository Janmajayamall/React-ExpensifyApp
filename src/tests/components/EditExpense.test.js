import React from 'react'
import {shallow} from 'enzyme';
import {EditExpensePage} from './../../components/EditExpense.js';
import { wrap } from 'module';
import expenses from './../fixtures/expenses'


let editExpense, startRemoveExpense, wrapper, history;

beforeEach(()=>{
    editExpense = jest.fn();
    startRemoveExpense = jest.fn();
    history = {push: jest.fn()}
    wrapper = shallow(<EditExpensePage editExpense = {editExpense} startRemoveExpense = {startRemoveExpense} history = {history} expense = {expenses[0]} />);
});

test('it should render EditExpensePage properly', ()=>{
    expect(wrapper).toMatchSnapshot();
}); 

test('should handle editExpense properly', ()=>{

    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
});

test('should remove expense properly', ()=>{

    wrapper.find('button').prop('onClick')();
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startRemoveExpense).toHaveBeenLastCalledWith(expenses[0].id);
});


