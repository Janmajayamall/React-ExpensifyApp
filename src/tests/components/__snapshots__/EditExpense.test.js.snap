// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should render EditExpensePage properly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 2,
          "createdAt": -345600000,
          "description": "gum",
          "id": 1,
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
</div>
`;
